bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/AsyncModbusFailure.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.io.transport.modbus;
+
+import java.util.Objects;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * Encapsulates result of modbus read operations
+ *
+ * @author Nagy Attila Gabor - Initial contribution
+ */
+@NonNullByDefault
+public class AsyncModbusFailure<R> {
```

Comment:
Could we replace `ModbusFailureCallback<R>` with `BiConsumer<R, Exception>`?
Similarly, replace `ModbusReadCallback` with `BiConsumer<R, ModbusData>` and `ModbusWriteCallback` with `BiConsumer<R, ModbusResponse>`? 

`ModbusData` would be simple DTO similar to `AsyncModbusReadResult`, containing `Optional<BitArray>`and `Optional<ModbusRegisterArray>`.

To work-around type erasure issues (see my separate comment), we might need

- `ModbusReadFailureCallback extends BiConsumer<ModbusReadRequestBlueprint, Exception>`
- `ModbusWriteFailureCallback extends BiConsumer<ModbusWriteRequestBlueprint, Exception>`
- `ModbusReadSuccessCallback extends BiConsumer<ModbusReadRequestBlueprint, ModbusData>` 
- `ModbusWriteSuccessCallback extends BiConsumer<ModbusWriteRequestBlueprint, ModbusData>` 

What do you think?

Reply:
I don't really see the benefit of BiConsumer here. It would be ok, if we want to avoid define our custom class, but once we extend it it does not have many added value I think.

Also I've seen that with the TaskWithEndpoint you moved to the direction of having more generics classes. These callback classes do not really add to the functionality by separating each of them. They could be more generics, to have fewer classes in the api.

I think the type erasure is not really such a big deal, because if you face it it might be a sign that you should split the concerns. Usually you should not implement these interfaces on a class, but they are here to create lambdas IMHO.

Reply:
I agree with your point of "not trying to avoid the type erasure issue ". As noted, lambdas are the intended way to use the callbacks 

With this in mind, there is no need to extend from BiConsumers, is there? We could use them directly as is, replacing the need to have custom Functional interfaces.

I'm fine either way, so leaving this up to you.